### PR TITLE
Feature/pelagos 4314 voter for files api

### PR DESCRIPTION
**Pull request by fightingtexasaggie**
not sure how we're going to tie in the API actions just yet. I can tie in DELETE for now. Also, is it too early to add a READ check to the voter?

### CODE REVIEW

File: src/Security/Voter/PelagosEntityVoter.php [hunk 1]
```diff
@@ -29,6 +29,7 @@ abstract class PelagosEntityVoter extends Voter
 {
     const CAN_CREATE = 'CAN_CREATE';
     const CAN_EDIT = 'CAN_EDIT';
+    const CAN_READ   = 'CAN_READ';
```
Comment (mickel1138) [inline src/Security/Voter/PelagosEntityVoter.php:32]: Everyone can always read, we probably don't have to vote on that.

File: src/Security/Voter/FileVoter.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * The Symfony calling security framework calls supports before calling voteOnAttribute.
+     *
+     * @param string         $attribute The action to be considered.
+     * @param mixed          $subject   A Dataset.
+     * @param TokenInterface $token     A security token containing user authentication information.
+     *
+     * @return boolean True If the user has one of the target roles for any of the subject's DataRepositories.
+     */
+    protected function voteOnAttribute($attribute, $subject, TokenInterface $token) //phpcs:ignore
```
Comment (praneethpr) [inline src/Security/Voter/FileVoter.php:54]: I think this voter should be the same as DatasetSubmissionVoter

File: src/Security/Voter/PelagosEntityVoter.php (hunk 1 above)
Comment (fightingtexasaggie) [reply src/Security/Voter/PelagosEntityVoter.php:32]: I was thinking of restricted mode later down the line, but yeah, we can implement that when we need to.

File: src/Security/Voter/FileVoter.php (hunk 2 above)
Comment (fightingtexasaggie) [reply src/Security/Voter/FileVoter.php:72]: Do you think it should consider the state of the submission as well?

**Comment by fightingtexasaggie** [comment]
Yesterday we put our heads together and realized that the approach to use a querybuilder within a file voter wasn't the best approach. I added commits to add a RRAR relationship from Fileset back to DatasetSubmission, but discussing this, we decided it would result in a circular reference making either object undeletable. 

By using some of the Doctrine tools, we determined what seemed to be the correct way of establishing these bi-directional OneToOne relationships, without the RRAR construct. What we still have left to determine is whether or not a separate voter is needed, or if the DatasetSubmission voter could be extended with additional attributes to handle File. The Api works on objects almost stand-alone, imho, and I don't think it will be able to relate File/Fileset to DatasetSubmission simply by virtue of it being in the same voter. I'm under the impression we will need a separate File voter that can be referenced in the File API. This voter will need the bi-directional relationship to figure out it's related DatasetSubmission (so that research groups can be found, etc)

File: src/Entity/File.php [hunk 3]
```diff
@@ -120,6 +120,11 @@ class File extends Entity
      */
     protected $status = self::FILE_NEW;
 
+    /**
+     * ORM\OneToOne(targetEntity=DatasetSubmission::class, mappedBy="File", cascade={"persist", "remove"})
+     */
+    private $datasetSubmission;
```
Comment (mickel1138) [inline src/Entity/File.php:126]: File should NOT have relationship with DatasetSubmission, it SHOULD have a relationship with FileSet

Comment (fightingtexasaggie) [reply src/Entity/File.php]: This was an overlooked mistake. I'd already updated Fileset but left this in from earlier.

**Review by mickel1138 — APPROVED** [review]

**Review by praneethpr — APPROVED** [review]

**Pull request merged**